Ingest managed domains for ResourceLoadStatistics and loosen restrictions for them.

pascoej opened the pull request:
#### 3a69dcff7becc04c05bd108b7cdd53f69426d056
<pre>
Ingest managed domains for ResourceLoadStatistics and loosen restrictions for them.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246290">https://bugs.webkit.org/show_bug.cgi?id=246290</a>
rdar://97344886

Reviewed by John Wilander.

This patch starts to ingest a small set of managed domains that are specically allowed
to use certain features normally disabled by resource load statistics. This patch ingests
the set on macOS and other Apple platforms.

The behavior change with regards to ResourceLoadStatistics is similar to what we
do for app-bound domains, but not exactly. These domains are exempted from certain deletion and
blocking policies.

* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-cookie-blocking-between-each-other-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-cookie-blocking-between-each-other.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion-database-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion-database.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion.html: Added.
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::shouldBlockCookies const):
(WebCore::NetworkStorageSession::setManagedDomains):
(WebCore::NetworkStorageSession::resetManagedDomains):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::setManagedDomains):
(WebKit::ResourceLoadStatisticsStore::shouldExemptFromWebsiteDataDeletion const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setAppBoundDomains):
(WebKit::WebResourceLoadStatisticsStore::setManagedDomains):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setManagedDomainsForResourceLoadStatistics):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetManagedDomainsForTesting):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setManagedDomainsForResourceLoadStatistics):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::managedDomainQueue):
(WebKit::WebsiteDataStore::platformInitialize):
(WebKit::managedDomains):
(WebKit::WebsiteDataStore::initializeManagedDomains):
(WebKit::WebsiteDataStore::ensureManagedDomains const):
(WebKit::WebsiteDataStore::getManagedDomains const):
(WebKit::WebsiteDataStore::managedDomainsIfInitialized):
(WebKit::WebsiteDataStore::setManagedDomainsForTesting):
(WebKit::WebsiteDataStore::reinitializeManagedDomains):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::forwardManagedDomainsToITPIfInitialized):
(WebKit::WebsiteDataStore::setManagedDomainsForITP):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setManagedDomains):
(WTR::TestRunner::didSetManagedDomainsCallback):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::ManagedDomainsCallbackContext::ManagedDomainsCallbackContext):
(WTR::didSetManagedDomainsCallback):
(WTR::TestController::setManagedDomains):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didSetManagedDomains):
* Tools/WebKitTestRunner/TestInvocation.h:

Canonical link: <a href="https://commits.webkit.org/256119@main">https://commits.webkit.org/256119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42d46a755cd19611a7f798fa7f38370c16a4faf5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3920 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/27641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/104375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98762 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3992 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/32101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100433 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81121 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/87072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/27641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/85926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38511 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/27641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81138 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/27641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/27895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40269 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83810 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2017 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42243 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/27641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/18930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -155,6 +155,10 @@
 
+#if !defined(MANAGED_DOMAINS)
+#define ENABLE_MANAGED_DOMAINS 0
+#endif
```
This has a mistake in it, `!defined(MANAGED_DOMAINS)` without the `ENABLE` prefix.

I wonder if we need this. Don’t features default to 0 even if we don’t do anything? I’m not sure we ever need to define something to 0 like this, even though we have lots of existing cases where we do.

darinadler commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -134,6 +134,7 @@ bool NetworkStorageSession::shouldBlockCookies(const ResourceRequest& request, s
     
 bool NetworkStorageSession::shouldBlockCookies(const URL& firstPartyForCookies, const URL& resource, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) const
 {
+
```
Please don’t add this blank line.

darinadler commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -666,6 +666,7 @@ void WebResourceLoadStatisticsStore::setAppBoundDomains(HashSet<RegistrableDomai
     auto domainsCopy = crossThreadCopy(domains);
 
     if (m_networkSession) {
+
```
Please don’t add this blank line.

darinadler commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -682,6 +683,35 @@ void WebResourceLoadStatisticsStore::setAppBoundDomains(HashSet<RegistrableDomai
+{
+    ASSERT(RunLoop::isMain());
+
+    if (isEphemeral() || domains.isEmpty()) {
```
Why is it OK to silently do nothing if someone passes in an empty set? What if the set in the statistics store is not already empty?

pascoej replied:
We don't want to change the cookie blocking mode if it's an empty set. The set being empty implies there's no managed websites.

darinadler replied:
Sure, but what if someone first set it to a non-empty set and is now setting it back to the empty set. Does the code handle that case properly? Or is that a forbidden sequence of calls?

darinadler commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -266,6 +266,9 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
 #endif
+#if ENABLE(MANAGED_DOMAINS)
+    void setManagedDomainsForResourceLoadStatistics(PAL::SessionID, HashSet<WebCore::RegistrableDomain>&&, CompletionHandler<void()>&&);
+#endif // ENABLE(MANAGED_DOMAINS)
```
These end comments are not great when the `#if` statement just covers a few lines. I see those throughout the patch, but not in the pre-existing code nearby.

darinadler commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -448,6 +459,13 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
 #endif
 
+#if ENABLE(MANAGED_DOMAINS)
+    static std::optional<HashSet<WebCore::RegistrableDomain>> managedDomainsIfInitialized();
```
Do we really want a function call that makes a copy of the set?

pascoej replied:
I think we do here. managedDomains() returns a reference to a static NeverDestroyed set, a similar pattern is used in appBoundDomainsIfInitialized().

darinadler replied:
That does not seem right. If we want to conditionally return a reference to the set, only if it’s already there, that would be a pointer! Not an optional copy. Unless there’s a reason we want to copy the entire set.

darinadler reviewed (changes requested):
Someone else should probably review, but I noticed at least one thing that should not land as-is.

johnwilander commented on Source/WebCore/platform/network/NetworkStorageSession.cpp:
```diff
@@ -159,13 +159,16 @@ bool NetworkStorageSession::shouldBlockCookies(const URL& firstPartyForCookies,
     case ThirdPartyCookieBlockingMode::AllExceptBetweenAppBoundDomains:
         return !shouldExemptDomainPairFromThirdPartyCookieBlocking(firstPartyDomain, resourceDomain);
+    case ThirdPartyCookieBlockingMode::AllExceptManagedDomains:
+        return !m_managedDomains.contains(firstPartyDomain) || !hasHadUserInteractionAsFirstParty(firstPartyDomain);
```
This is complex to understand. I suggest a helper variable and flipping the OR to an AND. Like this:
```
bool isManagedDomainWithUserInteraction = m_managedDomains.contains(firstPartyDomain) && hasHadUserInteractionAsFirstParty(firstPartyDomain);
return !isManagedDomainWithUserInteraction;
```

johnwilander commented on LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-cookie-blocking-between-each-other-expected.txt:
```diff
@@ -0,0 +1,33 @@
+Tests that managed domains are exempt from third-party cookie blocking between themselves.
```
I agree this should work but we also intend to exempt other 3p domains, right? Or is it just between the domains in the list?

johnwilander commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1278,6 +1278,15 @@ void NetworkProcessProxy::setAppBoundDomainsForResourceLoadStatistics(PAL::Sessi
 #endif
 
+#if ENABLE(MANAGED_DOMAINS)
+void NetworkProcessProxy::setManagedDomainsForResourceLoadStatistics(PAL::SessionID sessionID, const HashSet<RegistrableDomain>& appBoundDomains, CompletionHandler<void()>&& completionHandler)
```
When in bootstrap does this get set? I'm just trying to understand the order of things and runtime vs launch time.

pascoej replied:
`WebsiteDataStore::platformInitialize` -> `WebsiteDataStore::initializeManagedDomains` -> `WebsiteDataStore::forwardManagedDomainsToITPIfInitialized` -> `WebsiteDataStore::setManagedDomainsForITP` -> `setManagedDomainsForResourceLoadStatistics` (this)

johnwilander commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -699,6 +712,120 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+        if (hasInitializedManagedDomains && forceReinitialization != ForceReinitialization::Yes)
+            return;
+        static const auto maxManagedDomainCount = 10;
+        NSArray<NSString *> *crossSiteTrackingAllowed;
```
I think a better name is crossSiteTrackingPreventionRelaxed. It's not that we ever *allow* cross-site tracking.

johnwilander reviewed (changes requested):
Good to go as long as you make a few suggested changes and make sure our tests cover all that should be allowed under this setting.

johnwilander commented on Source/WebCore/PAL/pal/spi/ios/ManagedConfigurationSPI.h:
```diff
@@ -33,6 +33,10 @@ WTF_EXTERN_C_BEGIN
 #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 140000
 #import <ManagedConfiguration/ManagedConfiguration.h>
+@interface MCProfileConnection ()
+- (NSArray<NSString *> *)crossSiteTrackingAllowedDomains;
```
I don't think we should call this cross-site tracking allowed. We don't allow cross-site tracking. I suggest `crossSiteTrackingPreventionDisabledDomains`. You can go with `trackingPreventionDisabledDomains` if you think it gets too wordy.

pascoej replied:
I can't easily change the actual MDM names in this patch, but I will use better names within WenKit and make an issue regarding the MDM strings.

johnwilander commented on Source/WebCore/PAL/pal/spi/ios/ManagedConfigurationSPI.h:
```diff
@@ -62,6 +66,7 @@ typedef enum MCRestrictedBoolType {
 + (MCProfileConnection *)sharedConnection;
 - (MCRestrictedBoolType)effectiveBoolValueForSetting:(NSString *)feature;
 - (BOOL)isURLManaged:(NSURL *)url;
+- (NSArray<NSString *> *)crossSiteTrackingAllowedDomains;
```
Ditto.

johnwilander commented on Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
```diff
@@ -701,6 +701,31 @@ void WKWebsiteDataStoreSetAppBoundDomainsForTesting(WKArrayRef originURLsRef, vo
 #endif
 }
 
+void WKWebsiteDataStoreSetManagedDomainsForTesting(WKArrayRef originURLsRef, void* context, WKWebsiteDataStoreSetAppBoundDomainsForTestingFunction completionHandler)
```
We usually introduce a unique one of these per purpose: `WKWebsiteDataStoreSetAppBoundDomainsForTestingFunction`. Was there a reason to reuse the app-bound domains one?

johnwilander commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -699,6 +712,119 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+}
+
+NSString *kManagedSitesIdentifier = @"com.apple.mail-shared";
+NSString *kcrossSiteTrackingRelaxedDomainsKey = @"crossSiteTrackingRelaxedDomains";
```
`crossSiteTrackingPreventionRelaxed` is better. We're relaxing the prevention. You can go with just `trackingPreventionRelaxed` if you think it gets too wordy.

johnwilander commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -699,6 +712,119 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+        if (hasInitializedManagedDomains && forceReinitialization != ForceReinitialization::Yes)
+            return;
+        static const auto maxManagedDomainCount = 10;
+        NSArray<NSString *> *crossSiteTrackingRelaxed;
```
Ditto.

johnwilander commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -699,6 +712,119 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+#endif
+        managedKeyExists = crossSiteTrackingRelaxed ? true : false;
+    
+        RunLoop::main().dispatch([forceReinitialization, crossSiteTrackingRelaxed = retainPtr(crossSiteTrackingRelaxed)] {
```
Ditto.

johnwilander reviewed (changes requested):
Please update the naming. Other than that, good to go.

pascoej committed:
Ingest managed domains for ResourceLoadStatistics and loosen restrictions for them.
https://bugs.webkit.org/show_bug.cgi?id=246290
rdar://97344886

Reviewed by John Wilander.

This patch starts to ingest a small set of managed domains that are specically allowed
to use certain features normally disabled by resource load statistics. This patch ingests
the set on macOS and other Apple platforms.

The behavior change with regards to ResourceLoadStatistics is similar to what we
do for app-bound domains, but not exactly. These domains are exempted from certain deletion and
blocking policies.

* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-cookie-blocking-between-each-other-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-cookie-blocking-between-each-other.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion-database-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion-database.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-exempt-from-website-data-deletion.html: Added.
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::shouldBlockCookies const):
(WebCore::NetworkStorageSession::setManagedDomains):
(WebCore::NetworkStorageSession::resetManagedDomains):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::setManagedDomains):
(WebKit::ResourceLoadStatisticsStore::shouldExemptFromWebsiteDataDeletion const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setAppBoundDomains):
(WebKit::WebResourceLoadStatisticsStore::setManagedDomains):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setManagedDomainsForResourceLoadStatistics):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetManagedDomainsForTesting):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setManagedDomainsForResourceLoadStatistics):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::managedDomainQueue):
(WebKit::WebsiteDataStore::platformInitialize):
(WebKit::managedDomains):
(WebKit::WebsiteDataStore::initializeManagedDomains):
(WebKit::WebsiteDataStore::ensureManagedDomains const):
(WebKit::WebsiteDataStore::getManagedDomains const):
(WebKit::WebsiteDataStore::managedDomainsIfInitialized):
(WebKit::WebsiteDataStore::setManagedDomainsForTesting):
(WebKit::WebsiteDataStore::reinitializeManagedDomains):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::forwardManagedDomainsToITPIfInitialized):
(WebKit::WebsiteDataStore::setManagedDomainsForITP):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setManagedDomains):
(WTR::TestRunner::didSetManagedDomainsCallback):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::ManagedDomainsCallbackContext::ManagedDomainsCallbackContext):
(WTR::didSetManagedDomainsCallback):
(WTR::TestController::setManagedDomains):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didSetManagedDomains):
* Tools/WebKitTestRunner/TestInvocation.h:

Canonical link: https://commits.webkit.org/256119@main
